### PR TITLE
First round of fixes for supporting python3

### PR DESCRIPTION
**Pull request by natoscott**
Primarily updates to print() statements and using non-relative module imports.

The python3 build still fails though - see below - it looks like there's a required constructor parameter somewhere that's not got a default (I can't see where though...?).  I'll have to leave this one with you.

> python3 setup.py test

running test
Searching for redis-py-cluster
Best match: redis-py-cluster 1.3.5
Processing redis_py_cluster-1.3.5-py3.6.egg

Using /source/git/RLTest/.eggs/redis_py_cluster-1.3.5-py3.6.egg
running egg_info
writing RLTest.egg-info/PKG-INFO
writing dependency_links to RLTest.egg-info/dependency_links.txt
writing entry points to RLTest.egg-info/entry_points.txt
writing requirements to RLTest.egg-info/requires.txt
writing top-level names to RLTest.egg-info/top_level.txt
reading manifest file 'RLTest.egg-info/SOURCES.txt'
writing manifest file 'RLTest.egg-info/SOURCES.txt'
running build_ext
Traceback (most recent call last):
  File "setup.py", line 14, in <module>
    '''
  File "/usr/lib/python3.6/site-packages/setuptools/__init__.py", line 129, in setup
    return distutils.core.setup(**attrs)
  File "/usr/lib64/python3.6/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/usr/lib64/python3.6/distutils/dist.py", line 955, in run_commands
    self.run_command(cmd)
  File "/usr/lib64/python3.6/distutils/dist.py", line 974, in run_command
    cmd_obj.run()
  File "/usr/lib/python3.6/site-packages/setuptools/command/test.py", line 226, in run
    self.run_tests()
  File "/usr/lib/python3.6/site-packages/setuptools/command/test.py", line 248, in run_tests
    exit=False,
  File "/usr/lib64/python3.6/unittest/main.py", line 94, in __init__
    self.parseArgs(argv)
  File "/usr/lib64/python3.6/unittest/main.py", line 118, in parseArgs
    self._do_discovery(argv[2:])
  File "/usr/lib64/python3.6/unittest/main.py", line 229, in _do_discovery
    self.test = loader.discover(self.start, self.pattern, self.top)
  File "/usr/lib64/python3.6/unittest/loader.py", line 341, in discover
    tests = list(self._find_tests(start_dir, pattern))
  File "/usr/lib64/python3.6/unittest/loader.py", line 398, in _find_tests
    full_path, pattern, namespace)
  File "/usr/lib64/python3.6/unittest/loader.py", line 475, in _find_test_path
    tests = self.loadTestsFromModule(package, pattern=pattern)
  File "/usr/lib/python3.6/site-packages/setuptools/command/test.py", line 52, in loadTestsFromModule
    tests.append(self.loadTestsFromName(submodule))
  File "/usr/lib64/python3.6/unittest/loader.py", line 204, in loadTestsFromName
    test = obj()
TypeError: __init__() missing 1 required positional argument: 'redisBinaryPath'

### CODE REVIEW

**Comment by mnunberg** [comment]
+1

**Comment by MeirShpilraien** [comment]
@natoscott pls sign the Contributor License Agreement and we will merge it, thanks.

**Comment by gkorland** [comment]
@natoscott sorry but to protect both sides legally you're required to accept the CLA. I can assure you we're not planing to re-license the project. More over since it's under BSD you'll always be able to go back to this point in time and fork it.

**Comment by natoscott** [comment]
> @natoscott sorry but to protect both sides legally you're required to accept the CLA.

@gkorland you understand the problem open source developers have in contributing to projects where RedisLabs holds exclusive copyright now, right?  People want to contribute with the confidence their changes will remain open source - recent licensing changes have unfortunately destroyed the trust and goodwill RedisLabs once had built up in the open source community.

> I can assure you we're not planing to re-license the project.

OK, great - then surely there is no issue with sharing copyright here?  Can we use a different CLA for this project, which removes that clause from the CLA.

> you'll always be able to [...] fork it.

It's my preference to collaborate on the project in an on-going mutually beneficial way, where trust is built and responsibility shared over time, and copyright is not vested in one individual or company only.  This means any proposed future license change would be the responsibility of all contributors, and for the good of the project as a whole.

cheers.

**Comment by natoscott** [comment]
@gkorland ah, that's great news - I really appreciate that, thanks.

@MeirShpilraien did you have any clues for me as to where I might find the remaining problematic init() call that references redisBinaryPath?  I've not been able to locate it, and I'm hoping this may be the last step to achieving python3 support.

| TypeError: init() missing 1 required positional argument: 'redisBinaryPath'

**Comment by mnunberg** [comment]
A nice pretty please would also be to use format strings :)

**Comment by MeirShpilraien** [comment]
@natoscott I just checkout you branch and it works fine for me on python3 (I needed to install redis, redis-py-cluster, and psutils but thats it). The only issue I got is that when running the example tests, they failed because the result from the redis client is a binary string which are not equals the unicode string
```
test_example.__init__:
	testExample
	assertion faild:	b'1' == 1	test_example.py:16
	Test Failed
	testExample1
	assertion faild:	b'1' == 1	test_example.py:21
	assertion faild:	True == False	test_example.py:22
	Test Failed
	testExample2
	assertion faild:	b'1' == 1	test_example.py:27
	Test Failed
```

It can be fixed using the decode() function on the binary string.

Regardless, We made changes to the framework and currently your branch can not be merge because of conflicts. Is it possible that you take a look and fix them?

Thanks

File: RLTest/__main__.py [hunk 1]
```diff
@@ -221,9 +51,127 @@ def __exit__(self, type, value, traceback):
 class RLTest:
 
     def __init__(self):
-
-        # adding the current path to sys.path for test import puspused
-        sys.path.append(os.getcwd())
+        parser = CustomArgumentParser(fromfile_prefix_chars=RLTest_CONFIG_FILE_PREFIX,
+                                      formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+                                      description='Test Framework for redis and redis module')
+
```
Comment (MeirShpilraien) [inline RLTest/__main__.py]: we got this argument definition outside of the __init__ function

File: RLTest/__main__.py [hunk 2]
```diff
@@ -1,20 +1,25 @@
...
+import subprocess
+import platform
 import shutil
 import inspect
 import unittest
 import time
 import shlex
-
-from RLTest.env import Env, TestAssertionFailure
+from RLTest.Env import Env
```
Comment (MeirShpilraien) [inline RLTest/__main__.py]: It should be `from RLTest import Env` thats how it works for me.

File: RLTest/__main__.py [hunk 3]
```diff
@@ -281,249 +215,201 @@ def __init__(self):
...
-            sys.exit(0)
+        self.currEnv = None
 
     def _convertArgsType(self):
         pass
 
-    def takeEnvDown(self, fullShutDown=False):
-        if not self.currEnv:
-            return
+    def _downloadEnterpriseBinaries(self):
```
Comment (MeirShpilraien) [inline RLTest/__main__.py]: this function was removed to another file

File: RLTest/__main__.py [hunk 4]
```diff
@@ -281,249 +215,201 @@ def __init__(self):
...
+        # TODO: Support centos that does not have dpkg command
+        args = ['dpkg', '-x', os.path.join(RLTest_WORKING_DIR, debFileName), RLTest_WORKING_DIR]
+        self.process = subprocess.Popen(args=args, stdout=sys.stdout, stderr=sys.stdout)
+        self.process.wait()
+        if self.process.poll() != 0:
+            raise Exception('failed to extract binaries to %s' % self.RLTest_WORKING_DIR)
+
+        print(Colors.Yellow('finished installing enterprise binaries'))
+
+    def _loadFileTests(self, module_name):
```
Comment (MeirShpilraien) [inline RLTest/__main__.py]: this function was removed to loader

File: RLTest/__main__.py [hunk 5]
```diff
@@ -281,249 +215,201 @@ def __init__(self):
...
+                self.tests.append(getattr(module, func))
+                continue
+            if self.args.test_name:
+                if func == self.args.test_name:
+                    self.tests.append(getattr(module, func))
+                    return
+            elif func.startswith('test') or func.startswith('Test'):
+                self.tests.append(getattr(module, func))
+
+    def _loadTests(self):
```
Comment (MeirShpilraien) [inline RLTest/__main__.py]: also was removed to loader

File: RLTest/__main__.py [hunk 6]
```diff
@@ -281,249 +215,201 @@ def __init__(self):
...
-                if not failures:
-                    print '\t\t' + Colors.Bred('Exception raised during test execution. See logs')
-                for failure in failures:
-                    print '\t\t' + failure
+        print(Colors.Bold('Test Took: %d sec' % (endTime - startTime)))
+        print(Colors.Bold('Total Tests Run: %d, Total Tests Failed: %d, Total Tests Passed: %d' % (done, len(self.testsFailed), done - len(self.testsFailed))))
+        if len(self.testsFailed) > 0:
+            print(Colors.Bold('Failed Tests Summary:'))
+            for testFaild in self.testsFailed:
+                print('\t' + Colors.Bold(testFaild.testNamePrintable))
```
Comment (MeirShpilraien) [inline RLTest/__main__.py]: failure handling was changed, actually `testNamePrintable` is not longer exists

**Comment by natoscott** [comment]
| The python installation works for me.

@MeirShpilraien I'm not 100% sure what you mean there, but the code definitely doesn't work with python3 (master branch cannot, it has syntax errors - code which was valid under python2 but which it not under python3 - and my commits fix many issues, but do not attempt to fix everything).

| I think a better approach here is to start again from master

I agree (and will do), git merge really made a mess of all the intervening changes & it seems my attempts to put it right have several holes.  I had been using 'python3 setup.py test' in RLTest, but evidently that's not a good way to to test it.  It certainly shows many of the python3 problems though, and my broken merge attempt at least resolved many of those.  I will try again.  :)

| Try running the test_example.py and see [...]

Can you post the command line you're using to run that example test script?  I'm not seeing it in the README.  Thanks!

Also, could you setup CircleCI on this repository?  Once we have python3 working, it would be good to get immediate feedback on regressions there, and CI should help us immensely with that.  Thanks!

**Comment by MeirShpilraien** [comment]
just installed it and it does working (tell me if I am doing it wrong)
```
~/work/RedisSearch/RLTest > python3 setup.py install
running install
Checking .pth file support in /usr/local/lib/python3.5/dist-packages/
/usr/bin/python3 -E -c pass
TEST PASSED: /usr/local/lib/python3.5/dist-packages/ appears to support .pth files
running bdist_egg
running egg_info
writing top-level names to RLTest.egg-info/top_level.txt
writing RLTest.egg-info/PKG-INFO
writing entry points to RLTest.egg-info/entry_points.txt
writing dependency_links to RLTest.egg-info/dependency_links.txt
writing requirements to RLTest.egg-info/requires.txt
reading manifest file 'RLTest.egg-info/SOURCES.txt'
writing manifest file 'RLTest.egg-info/SOURCES.txt'
installing library code to build/bdist.linux-x86_64/egg
running install_lib
running build_py
copying RLTest/redis_cluster.py -> build/lib/RLTest
copying RLTest/__main__.py -> build/lib/RLTest
copying RLTest/loader.py -> build/lib/RLTest
copying RLTest/redis_std.py -> build/lib/RLTest
copying RLTest/env.py -> build/lib/RLTest
copying RLTest/Enterprise/Dmc.py -> build/lib/RLTest/Enterprise
copying RLTest/Enterprise/EnterpriseClusterEnv.py -> build/lib/RLTest/Enterprise
copying RLTest/Enterprise/binaryrepo.py -> build/lib/RLTest/Enterprise
copying RLTest/Enterprise/__init__.py -> build/lib/RLTest/Enterprise
copying RLTest/Enterprise/CcsMock.py -> build/lib/RLTest/Enterprise
creating build/bdist.linux-x86_64/egg
creating build/bdist.linux-x86_64/egg/RLTest
copying build/lib/RLTest/redis_cluster.py -> build/bdist.linux-x86_64/egg/RLTest
copying build/lib/RLTest/debuggers.py -> build/bdist.linux-x86_64/egg/RLTest
copying build/lib/RLTest/__init__.py -> build/bdist.linux-x86_64/egg/RLTest
copying build/lib/RLTest/__main__.py -> build/bdist.linux-x86_64/egg/RLTest
copying build/lib/RLTest/loader.py -> build/bdist.linux-x86_64/egg/RLTest
creating build/bdist.linux-x86_64/egg/RLTest/Enterprise
copying build/lib/RLTest/Enterprise/Dmc.py -> build/bdist.linux-x86_64/egg/RLTest/Enterprise
copying build/lib/RLTest/Enterprise/EnterpriseClusterEnv.py -> build/bdist.linux-x86_64/egg/RLTest/Enterprise
copying build/lib/RLTest/Enterprise/binaryrepo.py -> build/bdist.linux-x86_64/egg/RLTest/Enterprise
copying build/lib/RLTest/Enterprise/__init__.py -> build/bdist.linux-x86_64/egg/RLTest/Enterprise
copying build/lib/RLTest/Enterprise/CcsMock.py -> build/bdist.linux-x86_64/egg/RLTest/Enterprise
copying build/lib/RLTest/utils.py -> build/bdist.linux-x86_64/egg/RLTest
copying build/lib/RLTest/redis_std.py -> build/bdist.linux-x86_64/egg/RLTest
copying build/lib/RLTest/OssClusterEnv.py -> build/bdist.linux-x86_64/egg/RLTest
copying build/lib/RLTest/Env.py -> build/bdist.linux-x86_64/egg/RLTest
copying build/lib/RLTest/env.py -> build/bdist.linux-x86_64/egg/RLTest
copying build/lib/RLTest/OssEnv.py -> build/bdist.linux-x86_64/egg/RLTest
byte-compiling build/bdist.linux-x86_64/egg/RLTest/redis_cluster.py to redis_cluster.cpython-35.pyc
byte-compiling build/bdist.linux-x86_64/egg/RLTest/debuggers.py to debuggers.cpython-35.pyc
byte-compiling build/bdist.linux-x86_64/egg/RLTest/__init__.py to __init__.cpython-35.pyc
byte-compiling build/bdist.linux-x86_64/egg/RLTest/__main__.py to __main__.cpython-35.pyc
byte-compiling build/bdist.linux-x86_64/egg/RLTest/loader.py to loader.cpython-35.pyc
byte-compiling build/bdist.linux-x86_64/egg/RLTest/Enterprise/Dmc.py to Dmc.cpython-35.pyc
byte-compiling build/bdist.linux-x86_64/egg/RLTest/Enterprise/EnterpriseClusterEnv.py to EnterpriseClusterEnv.cpython-35.pyc
byte-compiling build/bdist.linux-x86_64/egg/RLTest/Enterprise/binaryrepo.py to binaryrepo.cpython-35.pyc
byte-compiling build/bdist.linux-x86_64/egg/RLTest/Enterprise/__init__.py to __init__.cpython-35.pyc
byte-compiling build/bdist.linux-x86_64/egg/RLTest/Enterprise/CcsMock.py to CcsMock.cpython-35.pyc
byte-compiling build/bdist.linux-x86_64/egg/RLTest/utils.py to utils.cpython-35.pyc
byte-compiling build/bdist.linux-x86_64/egg/RLTest/redis_std.py to redis_std.cpython-35.pyc
byte-compiling build/bdist.linux-x86_64/egg/RLTest/OssClusterEnv.py to OssClusterEnv.cpython-35.pyc
byte-compiling build/bdist.linux-x86_64/egg/RLTest/Env.py to Env.cpython-35.pyc
byte-compiling build/bdist.linux-x86_64/egg/RLTest/env.py to env.cpython-35.pyc
byte-compiling build/bdist.linux-x86_64/egg/RLTest/OssEnv.py to OssEnv.cpython-35.pyc
creating build/bdist.linux-x86_64/egg/EGG-INFO
copying RLTest.egg-info/PKG-INFO -> build/bdist.linux-x86_64/egg/EGG-INFO
copying RLTest.egg-info/SOURCES.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
copying RLTest.egg-info/dependency_links.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
copying RLTest.egg-info/entry_points.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
copying RLTest.egg-info/requires.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
copying RLTest.egg-info/top_level.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
zip_safe flag not set; analyzing archive contents...
creating 'dist/RLTest-0.1-py3.5.egg' and adding 'build/bdist.linux-x86_64/egg' to it
removing 'build/bdist.linux-x86_64/egg' (and everything under it)
Processing RLTest-0.1-py3.5.egg
Removing /usr/local/lib/python3.5/dist-packages/RLTest-0.1-py3.5.egg
Copying RLTest-0.1-py3.5.egg to /usr/local/lib/python3.5/dist-packages
RLTest 0.1 is already the active version in easy-install.pth
Installing RLTest script to /usr/local/bin

Installed /usr/local/lib/python3.5/dist-packages/RLTest-0.1-py3.5.egg
Processing dependencies for RLTest==0.1
Searching for psutil==5.4.7
Best match: psutil 5.4.7
Adding psutil 5.4.7 to easy-install.pth file

Using /home/meir/.local/lib/python3.5/site-packages
Searching for redis-py-cluster==1.3.5
Best match: redis-py-cluster 1.3.5
redis-py-cluster 1.3.5 is already the active version in easy-install.pth

Using /usr/local/lib/python3.5/dist-packages
Searching for redis==2.10.6
Best match: redis 2.10.6
Adding redis 2.10.6 to easy-install.pth file

Using /home/meir/.local/lib/python3.5/site-packages
Finished processing dependencies for RLTest==0.1
```

You can run the test by running `RLTest` in git root directory (I will add it to the readme) notice that some test failed in order to demonstrate failure (its an example of how to write tests on this framework, we currently do not have unitests the check the framework itself).

After I will add unitests that check the framework itself I will add circleCI to the repo.

**Comment by natoscott** [comment]
> just installed it and it does working (tell me if I am doing it wrong)

Not wrong - the 'build' and 'install' setup.py options work for me too,
but you should see it fail at the next step ...

> ~/work/RedisSearch/RLTest > python3 setup.py install

```
python3 setup.py test
```

> You can run the test by running `RLTest` in git root directory (I will add it to the readme) notice that

Aha - OK, now we're on the same page I think.  I tried that and with python3 this is the failure I see (similar sorts of issues as with 'setup.py test' above - the obvious python3 stuff like print() anyway)

```
$ python3 RLTest
Traceback (most recent call last):
  File "/usr/lib64/python3.6/runpy.py", line 185, in _run_module_as_main
    mod_name, mod_spec, code = _get_main_module_details(_Error)
  File "/usr/lib64/python3.6/runpy.py", line 219, in _get_main_module_details
    return _get_module_details(main_name)
  File "/usr/lib64/python3.6/runpy.py", line 153, in _get_module_details
    code = loader.get_code(mod_name)
  File "<frozen importlib._bootstrap_external>", line 781, in get_code
  File "<frozen importlib._bootstrap_external>", line 741, in source_to_code
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "RLTest/__main__.py", line 63
    print 'print'
                ^
SyntaxError: Missing parentheses in call to 'print'. Did you mean print('print')?
$ python3 --version
Python 3.6.6
```

> After I will add unitests that check the framework itself I will add circleCI to the repo.

Good stuff - thanks @MeirShpilraien 

I'll throw away that merge, and work through the python3 issues I see from the top again, and send a new PR.

**Comment by natoscott** [comment]
This PR has been replaced by https://github.com/RedisLabsModules/RLTest/pull/18 which incorporates all feedback and notes here, and also resolves the string/bytes decoding issues transparently.

**Pull request closed without merging**